jlm/llvm/opt/CommonNodeElimination.cpp

```diff
@@ -0,0 +1,1017 @@
+    auto nextSet = sets_.size();
+    auto [it, added] = congruenceSetMapping_.try_emplace(&leader, nextSet);
+
+    if (!added)
+    {
+      // If the leader already has its own set, we are done
+      if (sets_[it->second].leader == &leader)
+      {
+        return it->second;
+      }
+
+      // Remove the output from the congruence set it is following
+      sets_[it->second].followers.Remove(&leader);
+      it->second = nextSet;
+    }
+
+    // Create the new set, lead by \p leader
+    sets_.emplace_back(CongruenceSet{ &leader, {} });
+    return nextSet;
+  }
+
+  /**
+   * Gets the leader of the congruence set with the given \p index.
+   * @param index the index of the congruence set
+   * @return the output that is the leader of the set
+   */
+  const rvsdg::Output &
+  getLeader(Index index) const
+  {
+    JLM_ASSERT(index < sets_.size());
+    return *sets_[index].leader;
+  }
+
+  /**
+   * Makes the given \p output a follower of the congruence set with the given \p index.
+   * If \p output already follows the congruence set, this is a no-op
+   * If \p output already follows a different congruence set, it is moved.
+   * \p output can not be the leader of a congruence set.
+   * @param index the index of the congruence set to follow
+   * @param follower the output that should become a follower
+   */
+  void
+  addFollower(Index index, const rvsdg::Output & follower)
+  {
+    JLM_ASSERT(index < sets_.size());
+
+    const bool newFollower = sets_[index].followers.insert(&follower);
+
+    // If the follower is already following the correct set, do nothing
+    if (!newFollower)
+      return;
+
+    const auto [it, added] = congruenceSetMapping_.try_emplace(&follower, index);
+
+    // If the follower already belonged to a congruence set, remove it from the old set
+    if (!added)
+    {
+      JLM_ASSERT(it->second != index);
+
+      if (sets_[it->second].leader == &follower)
+        throw std::logic_error("Cannot turn a leader into a follower");
+
+      const bool removed = sets_[it->second].followers.Remove(&follower);
+      JLM_ASSERT(removed);
+
+      it->second = index;
+    }
+  }
+
+  /**
+   * Gets the set of followers in the congruence set with the given \p index.
+   * @param index the index of the congruence set
+   * @return the followers
+   */
+  const util::HashSet<const rvsdg::Output *> &
+  getFollowers(Index index) const
+  {
+    JLM_ASSERT(index < sets_.size());
+    return sets_[index].followers;
+  }
+
+private:
+  // The list of congruence sets
+  std::vector<CongruenceSet> sets_;
+  // A mapping from T to the congruence set it belongs to, either as leader or follower
+  std::unordered_map<const rvsdg::Output *, Index> congruenceSetMapping_;
+};
+
+class CommonNodeElimination::Context final
+{
+
+public:
+  [[nodiscard]] CongruenceSets &
+  getCongruenceSets()
+  {
+    return congruenceSets_;
+  }
+
+private:
+  CongruenceSets congruenceSets_;
+};
```

Comment:
This indirection feels a bit useless with only a single member. Could the CongruenceSets class not just be renamed into context and we skip this indirection?

Reply:
done